report-service/src/main/java/gov/cdc/dataingestion/authservice/integration/service/AuthService.java

```diff
@@ -0,0 +1,180 @@
+    private String nbsPassword;
+
+    @Value("${auth.sign-on-url}")
+    private String signOnUrl;
+
+    @Value("${auth.token-url}")
+    private String tokenUrl;
+
+    @Value("${auth.salt-for-algorithm}")
+    private String saltForAlgorithm;
+
+    private String token;
+
+    private static final String AUTH_ELR_CLAIM = "ELR Importer";
+    private static final String AUTH_ECR_CLAIM = "ECR Importer";
+
+    private String encodedSignOnUrl;
+
+
+    @PostConstruct
+    public void generateAuthTokenDuringStartup() {
+        encodedSignOnUrl = getSignOnUrl();
+        generateToken(encodedSignOnUrl);
+
+        String authRoleName = getAuthRoleClaim(token);
+        if(authRoleName == null || authRoleName.isEmpty()) {
+            logger.error("Auth roles not defined, nothing to authorize.");
+            return;
+        }
+
+        logger.info("Auth role claim is: " + authRoleName);
+
+        boolean isUserAllowedToLoadElrData = authRoleName.contains(AUTH_ELR_CLAIM);
+        boolean isUserAllowedToLoadEcrData = authRoleName.contains(AUTH_ECR_CLAIM);
+
+        logger.info("Is user allowed to load ELR data: " + isUserAllowedToLoadElrData);
+        logger.info("Is user allowed to load ECR data: " + isUserAllowedToLoadEcrData);
+    }
+
+    // Need to generate token before it's expiry (1 hour) and that is
+    // why this method is scheduled every 55 minutes
+    @Scheduled(fixedRate = 55*60*1000)
+    public void generateAuthTokenScheduled() {
+        generateToken(tokenUrl);
+    }
+
+    private void generateToken(String url) {
+        try {
+            CloseableHttpClient httpClient = buildHttpClient();
+            if(httpClient != null) {
+                CloseableHttpResponse response;
+                if(url.contains("token")) {
+                    HttpGet getRequest = new HttpGet(url);
+                    getRequest.addHeader("Auth-Token", token);
+                    response = httpClient.execute(getRequest);
+                }
+                else {
+                    HttpPost postRequest = new HttpPost(url);
+                    response = httpClient.execute(postRequest);
+                }
+                int statusCode = response.getStatusLine().getStatusCode();
+
+                // TODO: Lookup the right code for the token expiration error.
+                //  Need to create a story to add this code logic in the Auth service
+                if(statusCode == 503) {
+                    logger.info("Token expired. Signing on again...");
+                    generateToken(encodedSignOnUrl);
+                }
+
+                HttpEntity httpEntity = response.getEntity();
+                ObjectMapper mapper = new ObjectMapper();
+                JsonNode node = mapper.readTree(EntityUtils.toString(httpEntity));
+                token = node.get("token").asText();
+            }
+            else {
+                logger.error("Http client returned as null.");
+            }
+        } catch (Exception e) {
```

Comment:
Probably better if we could add custom exception here. Ex: DIAuthenticationException or something similar

Reply:
fixed